### PR TITLE
Implement more general cv mechanism

### LINKED ISSUE #295: Implement more general cross validation mechanism
WHAT: At the moment, the model selection scenarios are relatively fixed. There are several foreseen use cases which are not covered. Mainly the train/test/validation spilt is quite fixed. To make more training startegies possible we want to make this procces more flexibible and confguriable. 

For instance:
* A cross site model where we want the same sites not to be both in train and test
* We might need more complicated train/val test splitting strategies.

Having this possibility is required to train fussion where that combine different models that are trained on the same data.

HOW:
- It should be possible to have a CV base object which can be overloaded to specify more complicated strategies. For instance,   in sklearn, it is possible to create custom cross validation objects, which are then passed to a Metamodel which deals with the 
  training process.
- Implement  a similar object in openSTEF that takes care of the train/test/vailadtion splitting strategy

TEST:
- Splitting of train and validation data is handled by a cross validator object
- "classic" strategy is pressent that implements the current strategy

### PR DESCRIPTION
**Pull request by AlexisMignon**
# Implement a more general cv mechanism.

## Configuration

We add two configuration fields to `PredictionJobDataClass`:

* `train_split_func`
* `backtest_split_fun`

Using a new data class `SplitFuncDataClass` with two fields:

* `function` either a string or a function
* `arguments` additional arguments to be provided to the function.

The train function is assumed to return a triplet of data frames (train, validation, test).
The backtest function is assumed to return an iterable on such triplets

## Modifications to `train_model` pipeline.
To be the most general, we factorized `train_pipeline_common` into subparts:
* feature application
* data splitting
* model training

The default data splitting function allows using a default function used a custom split function defined in the prediction job

This allows us to use those subfunctions in the backtest pipeline, which has been deeply modified.

## Modifications to `create_forecast_backtest`

We use the new subfunctions. We do not separate the case `n_folds == 1` any more. While we usually try to keep the code unchanged when new features are added, in this case we tried to reimplement the original cv mechanism using the new one. 

## Behaviour changes

The back test pipelines now returns the data sets with added features. Is that OK ?

Closes #295

### CODE REVIEW

**Comment by AlexisMignon** [comment]
Note that so far, we have tested that, when no configuration is provided, the existing tests still pass.

Before considering including the PR we should

* [ ] Validate the behaviour change of the backtest pipeline (datasets with features are now returned)
* [ ] check that the implemented default cv mechanism behaves as the original one.
* [ ] Implement unittest using configurations

File: openstef/pipeline/train_create_forecast_backtest.py [hunk 1]
```diff
@@ -32,150 +35,51 @@ def train_model_and_forecast_back_test(
     List[pd.DataFrame],
     List[pd.DataFrame],
 ]:
-    """Pipeline for a back test.
```
Comment (JanMaartenvanDoorn) [inline openstef/pipeline/train_create_forecast_backtest.py:35]: Here you remove the documentation of the `train_model_and_forecast_back_test` , Could you please update the documentation and add it to the updated function?

File: openstef/pipeline/train_create_forecast_backtest.py [hunk 2]
```diff
@@ -32,150 +35,51 @@ def train_model_and_forecast_back_test(
...
-    """
-    model, report, train_data, validation_data, test_data = train_pipeline_common(
-        pj,
-        modelspecs,
-        input_data,
-        training_horizons,
-        test_fraction=test_fraction,
-        backtest=True,
-        test_data_predefined=test_data,
+    pj, modelspecs, train_data, validation_data, test_data
```
Comment (JanMaartenvanDoorn) [inline openstef/pipeline/train_create_forecast_backtest.py:79]: Also here, please add some documentation (can be short) and type hints so it is clear how to use this function.

File: openstef/pipeline/train_create_forecast_backtest.py [hunk 3]
```diff
@@ -199,4 +103,32 @@ def train_model_and_forecast_test_core(
     forecast["realised"] = test_data.iloc[:, 0]
     forecast["horizon"] = test_data.iloc[:, -1]
 
-    return forecast, model, train_data, validation_data, test_data
+    return model, forecast
+
+
+def backtest_split_default(
+    input_data, n_folds, test_fraction=0.15, stratification_min_max=True
```
Comment (JanMaartenvanDoorn) [inline openstef/pipeline/train_create_forecast_backtest.py:110]: Please add documentation and typehints.

File: openstef/pipeline/train_model.py [hunk 4]
```diff
@@ -212,14 +212,74 @@ def train_pipeline_common(
             (empty data frame by default)
 
     Returns:
-        Tuple[RegressorMixin, Report, pd.DataFrame, pd.DataFrame, pd.DataFrame]: Trained model, report
-         train_data, validation_data and test_data
+        trained_model (RegressorMixin): the trained model
```
Comment (JanMaartenvanDoorn) [inline openstef/pipeline/train_model.py:215]: Nice updated documentation!

**Review by JanMaartenvanDoorn — CHANGES_REQUESTED** [review]
Nice way of making the cross validation more configurable! I have a few minor remarks about documentation and type hints, please check that alle methods have input typehints, output typehints and at least some documentation (does not have to be elaborate). Furthermore some thoughts on where to put the new back_test split function and renaming the new "pipeline_step" functions. I am curious about your thoughts.

File: openstef/pipeline/train_model.py [hunk 5]
```diff
@@ -212,14 +212,74 @@ def train_pipeline_common(
...
+        train_data (pd.DataFrame): The train data
+        validation_data (pd.DataFrame): The validation data
+        test_data (pd.DataFrame): The test data
 
     Raises:
         InputDataInsufficientError: when input data is insufficient.
         InputDataWrongColumnOrderError: when input data has a invalid column order.
 
     """
+    data_with_features = train_pipeline_compute_features(
```
[the remark this answers is not in the document]
Comment (AlexisMignon) [reply openstef/pipeline/train_model.py]: So I have renamed the functions with a train_pipeline_step_* prefix.

**Comment by AlexisMignon** [comment]
There are some coverage issues left which are treated in the PR linked to the `save_train_forecasts` PR.

**Review by JanMaartenvanDoorn — APPROVED** [review]
Good typehints and documentation!

**Comment by AlexisMignon** [comment]
@JanMaartenvanDoorn. Should we merge ?

**Comment by JanMaartenvanDoorn** [comment]
> @JanMaartenvanDoorn. Should we merge ?

I think this is fine, if you fix the merge conflicts you can merge it.

**Pull request merged**